clash-cores/src/Clash/Cores/Floating/Xilinx.hs

```diff
@@ -0,0 +1,104 @@
+named @floating_point@/.../@.tcl@ which needs to be executed in the Vivado
+project to create the proper entity. Simulation in Clash produces bit-identical
+results to synthesis.
+
+Most functions allow customization of the Xilinx IP. Valid combinations will
+need to be gleaned from, e.g., the Vivado wizard (IP Catalog -> Floating-point).
+All IP instantiated by this module always has the following properties:
+
+* Single precision
+* Non-blocking
+* No reset input
+* No optional output fields, no @TLAST@, no @TUSER@.
+* @TVALID@ on the inputs is always asserted, @TVALID@ on the output is ignored.
+Note that it would appear the Xilinx IP does not use these signals in
+non-blocking mode anyway.
+* 1 cycle per operation (meaning there need not be any dummy cycles between
+consecutive inputs)
+
+The latency of the IP is set through the delay argument of the 'DSignal'. Other
+customization is done through the 'FloatingConfig' argument of customizable
+functions.
+
+De-asserting the 'Enable' signal of a function will stall the whole pipeline.
+
+The Xilinx IP does not support calculations with subnormal numbers. All
+subnormal numbers are rounded to zero on both input and output. Note that this
+introduces a slight bias as the larger subnormal numbers are closer to the
+smallest normal number, but they are rounded to zero nonetheless!
+
+For each customizable operation, there also exists a function that uses the
+defaults. These functions use the settings from 'defFloatingC' and the maximum
+delay for the Xilinx IP with that configuration. That delay is also defined as
+a type variable for delay annotation in circuits.
+-}
+
+module Clash.Cores.Floating.Xilinx
+  ( -- * Instantiating IP
+    addFloat
+  , addFloat'
+  , AddFloatDefDelay
+    -- * Customizing IP
+  , FloatingConfig(..)
+  , defFloatingC
+  , FloatingArchOpt(..)
+  , FloatingDspUsage(..)
+  ) where
+
+import Clash.Prelude
+
+import GHC.Stack (HasCallStack, withFrozenCallStack)
+
+import Clash.Cores.Floating.Xilinx.Internal
+
+-- | Default customization options.
+--
+-- For those operations that support it, the default options are:
+--
+-- * Speed-optimized architecture ('SpeedArch').
+-- * Full DSP slice usage ('FullDspUsage').
+-- * Exponential operator does not use block memory.
+defFloatingC :: FloatingConfig
+defFloatingC = FloatingConfig
+  { floatingArchOpt = SpeedArch
+  , floatingDspUsage = FullDspUsage
+  , floatingBMemUsage = False
+  }
+
+-- | Floating point addition.
+addFloat
+  :: ( HiddenClock dom
+     , HiddenEnable dom
+     , KnownNat d
+     , HasCallStack
+     )
+  => FloatingConfig
+  -> DSignal dom n Float
+  -> DSignal dom n Float
+  -> DSignal dom (n + d) Float
+addFloat cfg = withFrozenCallStack $ hideEnable . hideClock $ addFloat# cfg
+{-# INLINE addFloat #-}
+
+-- | The default delay for floating point addition with default customization.
+type AddFloatDefDelay = 11
+
+-- | Floating point addition with default settings.
+addFloat'
```

Comment:
I'm not sure about the use of `'` here, mostly because it makes me think of strictness. In my own code I would tend to do something more like

```haskell
addFloatWith :: FloatingConfig -> ...
addFloatWith = ...

addFloat :: ...
addFloat =
  withFrozenCallStack $ addFloatWith defFloatingC
```

so it's clearer as someone reading just the haddock that one is configurable and one isn't

Reply:
I like the "with". The prime to me just suggests relation between prime and non-prime, but it's also a really small glyph, that's annoying.